packages/beacon-node/src/sync/utils/downloadByRoot.ts

```diff
@@ -0,0 +1,589 @@
+    peerIdStrings: cacheItem.peerIdStrings,
+  };
+}
+
+export async function fetchByRoot({
+  config,
+  network,
+  executionEngine,
+  peerIdStr,
+  blockRoot,
+  cacheItem,
+}: FetchByRootProps): Promise<FetchByRootResponses> {
+  let block: SignedBeaconBlock;
+  let blobSidecars: deneb.BlobSidecars | undefined;
+  let columnSidecars: fulu.DataColumnSidecars | undefined;
+
+  if (isPendingBlockInput(cacheItem)) {
+    if (cacheItem.blockInput.hasBlock()) {
+      block = cacheItem.blockInput.getBlock();
+    } else {
+      block = await fetchAndValidateBlock({
+        config,
+        network,
+        peerIdStr,
+        blockRoot,
+      });
+    }
+
+    const forkName = config.getForkName(block.message.slot);
+    if (!cacheItem.blockInput.hasAllData()) {
+      if (isBlockInputBlobs(cacheItem.blockInput)) {
+        blobSidecars = await fetchAndValidateBlobs({
+          config,
+          network,
+          executionEngine,
+          peerIdStr,
+          forkName: forkName as ForkPreFulu,
+          block: block as SignedBeaconBlock<ForkPostDeneb>,
+          blockRoot,
+          blobMeta: cacheItem.blockInput.getMissingBlobMeta(),
+        });
+      }
+      if (isBlockInputColumns(cacheItem.blockInput)) {
+        columnSidecars = await fetchAndValidateColumns({
+          config,
+          network,
+          executionEngine,
+          peerIdStr,
+          forkName: forkName as ForkPostFulu,
+          block: block as SignedBeaconBlock<ForkPostFulu>,
+          blockRoot,
+          columnMeta: cacheItem.blockInput.getMissingSampledColumnMeta(),
+        });
+      }
+    }
+  } else {
+    block = await fetchAndValidateBlock({
+      config,
+      network,
+      peerIdStr,
+      blockRoot,
+    });
+    const forkName = config.getForkName(block.message.slot);
+    if (isForkPostFulu(forkName)) {
+      columnSidecars = await fetchAndValidateColumns({
+        config,
+        network,
+        executionEngine,
+        peerIdStr,
+        forkName,
+        blockRoot,
+        block: block as SignedBeaconBlock<ForkPostFulu>,
+        columnMeta: {
+          missing: network.custodyConfig.sampledColumns,
+          versionedHashes: (block as SignedBeaconBlock<ForkPostFulu>).message.body.blobKzgCommitments.map((c) =>
+            kzgCommitmentToVersionedHash(c)
+          ),
+        },
+      });
+    } else if (isForkPostDeneb(forkName)) {
+      const commitments = (block as SignedBeaconBlock<ForkPostDeneb>).message.body.blobKzgCommitments;
+      const blobCount = commitments.length;
+      blobSidecars = await fetchAndValidateBlobs({
+        config,
+        network,
+        executionEngine,
+        peerIdStr,
+        forkName: forkName as ForkPreFulu,
+        blockRoot,
+        block: block as SignedBeaconBlock<ForkPostDeneb>,
+        blobMeta: Array.from({length: blobCount}, (_, i) => ({
+          index: i,
+          blockRoot,
+          versionedHash: kzgCommitmentToVersionedHash(commitments[i]),
```

Comment:
this is for `getBlobs()` predeneb, and it has the same concern/performance issue to what #8282 resolved:
- we call `getBlobs()` in a for loop of peers
- we may instead only call once we first seen a gossip block, in the [deneb spec](https://github.com/ethereum/consensus-specs/blob/master/specs/deneb/p2p-interface.md): `Honest nodes SHOULD
query `engine_getBlobsV1` as soon as they receive a valid gossip block that
contains data, and import the returned blobs.`
- `Buffer.fromHex()` is not great 

I suppose we need to resolve this issue on `unstable` first